### PR TITLE
Add linear learning rate scheduler

### PR DESCRIPTION
**Pull request by astral4**
## Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

## Related Issues/PRs

https://github.com/tracel-ai/burn/issues/1198

## Changes

This PR adds a linear learning rate scheduler to Burn, which didn't exist previously. This makes progress towards the issue linked above.

## Testing

I wrote tests to confirm that invalid config parameters are rejected and that the learning rate changes over iterations as expected. I successfully ran `run-checks.sh all`. I also checked the appearance of the new documentation via `cargo doc --open`.

### CODE REVIEW

File: crates/burn-core/src/lr_scheduler/linear.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+/// A linear learning rate scheduler.
+///
+/// See [LinearLrSchedulerConfig] for more information.
+#[derive(Clone, Copy, Debug)]
+pub struct LinearLrScheduler {
+    // The previous iteration's learning rate.
+    previous_lr: LearningRate,
+    // The amount that the learning rate changes by on each iteration.
+    step_size: LearningRate,
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/linear.rs:52]: Though `LearningRate` is currently an alias for `f64`, I wouldn't say the step size is a learning rate. Let's keep it primitive `step_size: f64`.

**Review by laggui — CHANGES_REQUESTED** [review]
Thanks for working on this! 🙏 

Overall, LGTM. Just a minor comment.

File: crates/burn-core/src/lr_scheduler/linear.rs (hunk 1 above)
Comment (astral4) [reply crates/burn-core/src/lr_scheduler/linear.rs]: Got it! I've changed it to `f64`.

File: crates/burn-core/src/lr_scheduler/linear.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            "Initial learning rate must be greater than 0 and at most 1"
+        );
+        assert!(
+            self.final_lr >= 0. && self.final_lr <= 1.,
+            "Final learning rate must be at least 0 and at most 1"
+        );
+
+        LinearLrScheduler {
+            previous_lr: self.initial_lr,
+            step_size: (self.final_lr - self.initial_lr) / self.num_iters as LearningRate,
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/linear.rs:38]: We should also use `as f64` here instead.
`step_size: (self.final_lr - self.initial_lr) / self.num_iters as f64`

File: crates/burn-core/src/lr_scheduler/linear.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    // The previous iteration's learning rate.
+    previous_lr: LearningRate,
+    // The amount that the learning rate changes by on each iteration.
+    step_size: f64,
+    // The number of iterations left before reaching the final learning rate.
+    remaining_iters: usize,
+}
+
+impl<B: Backend> LrScheduler<B> for LinearLrScheduler {
+    type Record = (LearningRate, LearningRate, usize);
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/linear.rs:58]: For consistency: `type Record = (LearningRate, f64, usize)` now that `step_size` is `f64`.

**Review by laggui — CHANGES_REQUESTED** [review]
Sorry I missed some comments in the last review since it is retroactive to the change of `step_size` to `f64`. The changes were kind of implicit but that's my fault 😅 

See my comments below, and then LGTM 🚀

**Comment by astral4** [comment]
Alright, it should be resolved now. Thanks!

**Review by laggui — APPROVED** [review]

**Review by nathanielsimard — APPROVED** [review]
